### PR TITLE
Ex1

### PR DESCRIPTION
**Pull request by YairAmar**
Exercise 1 is done and ready for a code review

### CODE REVIEW

File: ex1/LinearRegression.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+from Utils import *
```
Comment (yarinshaked) [inline ex1/LinearRegression.py:1]: file_name should be lower case, such as variables, e.g: utils, linear_regression,....

Comment (yarinshaked) [inline ex1/LinearRegression.py:1]: you should never import *, always specify what you need from the module

File: ex1/LinearRegression.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+class LinearRegression:
+    """
+    This class is made for training a model of linear regression
+    """
+
+    def __init__(self):
+        """
+        The constructor of the LinearRegression object
+        """
+        self.theta = np.zeros((2,1))
```
Comment (yarinshaked) [inline ex1/LinearRegression.py:13]: I suggest you to pass the shape values as an arguments to the constructor

File: ex1/LinearRegression.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    def fit(self, x, y, iterations=1500, learning_rate=0.01, save_cost=False,
+            save_theta=False, normalize_data=True):
+        """
```
Comment (yarinshaked) [inline ex1/LinearRegression.py:17]: use google docstring format

File: ex1/LinearRegression.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        y - target (expected output for the data)
+        iterations - number of gradient descent iterations, defaulted to be 1500
+        learning_rate -  defaulted to be 0.01
+        save_cost - if True, fit will return a list of the cost-function's values
+        save_theta - if True, fit will return a list of theta's values
+        """
+        if normalize_data:
+            feature_normalize(x)
+
+        self.theta = np.zeros((x.shape[1], 1))
```
Comment (yarinshaked) [inline ex1/LinearRegression.py:30]: 1 seems like a magic number, define it as const in consts.py file

Comment (yarinshaked) [inline ex1/LinearRegression.py:30]: why do you define self.theta again?

File: ex1/LinearRegression.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                theta_list.append(np.copy(self.theta))
+
+            y_hat = x @ self.theta
+            step = (y_hat - y).T @ x / m
+            self.theta -= learning_rate * step.T
+
+        return {"cost": cost_list, "theta": theta_list}
+
+    def fit_with_normal_eq(self, x, y):
+        """
```
Comment (yarinshaked) [inline ex1/LinearRegression.py:49]: again use google docstring

File: ex1/LinearRegression.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        The constructor of the LinearRegression object
+        """
+        self.theta = np.zeros((2,1))
+
+    def fit(self, x, y, iterations=1500, learning_rate=0.01, save_cost=False,
+            save_theta=False, normalize_data=True):
+        """
+        Trains the Linear-Regression model.
+        the model's weights are optimized using batch gradient descent
+        x1 - the input data
```
Comment (yarinshaked) [inline ex1/LinearRegression.py:20]: it should be the same name as the argument: "x"

File: ex1/LinearRegression.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            y_hat = x @ self.theta
+            step = (y_hat - y).T @ x / m
+            self.theta -= learning_rate * step.T
+
+        return {"cost": cost_list, "theta": theta_list}
+
+    def fit_with_normal_eq(self, x, y):
+        """
+        Solves the linear regression problem with a closed formula of normal equation
+        x1 - input data
```
Comment (yarinshaked) [inline ex1/LinearRegression.py:51]: again should be x

File: ex1/LinearRegression.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        y - target
+        returns theta, when it contains [theta_0,theta_1]
+        """
+        # used pseudo-inverse for stability
+        self.theta = np.linalg.pinv(x.T @ x) @ x.T @ y
+        return self.theta
+
+    def predict(self, x):
+        """
```
Comment (yarinshaked) [inline ex1/LinearRegression.py:60]: again google docstring

File: ex1/LinearRegression.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        x1 - input data
```
Comment (yarinshaked) [inline ex1/LinearRegression.py:61]: again should be x

File: ex1/Utils.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
+import numpy as np
```
Comment (yarinshaked) [inline ex1/Utils.py:1]: I like to import alphabetically but it is your call

File: ex1/Utils.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+import matplotlib.pyplot as plt
+
+
+def compute_cost(x, y, theta):
+    """
```
Comment (yarinshaked) [inline ex1/Utils.py:6]: again google

File: ex1/Utils.py [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    computing the cost function
+    x1 - input data, with a column of 1
```
Comment (yarinshaked) [inline ex1/Utils.py:8]: again x

File: ex1/LinearRegression.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        self.theta = np.zeros((x.shape[1], 1))
+        m = len(y)
+        cost_list = []
+        theta_list = []
+
+        for i in range(iterations):
+            # documenting the values of theta and the cost-function
+            if save_cost:
+                cost_list.append(compute_cost(x, y, self.theta))
+            if save_theta:
```
Comment (yarinshaked) [inline ex1/LinearRegression.py:39]: blank line between if statements

File: ex1/Utils.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+    returns the cost function's value
+    """
+    m = len(y)
+    y_hat = x @ theta
+    cost = ((y_hat - y).T @ (y_hat - y) / (2 * m)).item()
+    return cost
+
+
+def feature_normalize(x):
+    """
```
Comment (yarinshaked) [inline ex1/Utils.py:19]: again google

File: ex1/Utils.py [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    normalizes data to be ~ N(0,1)
+    x1 - input data
```
Comment (yarinshaked) [inline ex1/Utils.py:21]: again x

File: ex1/Utils.py [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+    """
+    normalize = lambda vec: (vec - np.mean(vec)) / np.std(vec)
```
Comment (yarinshaked) [inline ex1/Utils.py:23]: you can define an inner function within, instead of lambda expression

File: ex1/Utils.py [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+    for i in np.arange(x.shape[1] - 1) + 1:
```
Comment (yarinshaked) [inline ex1/Utils.py:24]: why not use range()?

File: ex1/Utils.py [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+        x[:, i] = normalize(x[:, i])
+
+
+def plot_model(x, y, theta):
+    """
```
Comment (yarinshaked) [inline ex1/Utils.py:29]: again google

File: ex1/Utils.py [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+    Works only for 1-d data
+    visualizes a trained regression model, while plotting the data as well
+    x1 - the input data
```
Comment (yarinshaked) [inline ex1/Utils.py:32]: again x

File: ex1/Utils.py [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+    plt.title("fitted model")
+    plt.grid()
+    t = np.linspace(np.min(x[:, -1]), np.max(x[:, -1]), 10000)
+    line = theta[0] + t * theta[1]
+    plt.plot(t, line, 'r')
+    plt.show()
+
+
+def visualize_optimization(x, y, doc_dict, f):
+    """
```
Comment (yarinshaked) [inline ex1/Utils.py:48]: again google

File: ex1/Utils.py [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+    Works only for 1-d data
+    creates a 3-d surface plot for the given cost-function
+    x1 - the input data
```
Comment (yarinshaked) [inline ex1/Utils.py:51]: again x

File: ex1/Utils.py [hunk 22]
```diff
@@ -0,0 +1,86 @@
...
+    y - target (expected output for the data)
+    doc_dict - doc-dictionary, output of the fit function in LinearRegression class
+    f - function that computes the wanted cost funtion
+    """
+    fig = plt.figure(figsize=(12, 12))
+    ax = fig.gca(projection='3d')
+
+    theta0s = np.arange(-10, 10, .5)
```
Comment (yarinshaked) [inline ex1/Utils.py:59]: you can find a better name

File: ex1/Utils.py [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+    theta1s = np.arange(-1, 8, .5)
```
Comment (yarinshaked) [inline ex1/Utils.py:60]: same here

File: ex1/Utils.py [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+    th0, th1, cost = [], [], []
```
Comment (yarinshaked) [inline ex1/Utils.py:61]: same here

File: ex1/Utils.py [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+            cost.append(f(x, y, tmp_theta))
+
+    ax.scatter(th0, th1, cost, c=np.abs(cost))
+    plt.xlabel(r'$\theta_0$', fontsize=30)
+    plt.ylabel(r'$\theta_1$', fontsize=30)
+    plt.title('Cost', fontsize=20)
+
+    theta_hist = doc_dict["theta"]
+    cost_hist = doc_dict["cost"]
+    ax.plot([th[0] for th in theta_hist], [th[1] for th in theta_hist], cost_hist, 'ro-')
```
Comment (yarinshaked) [inline ex1/Utils.py:76]: it would be more clearer if done outside the function invocation

File: ex1/Utils.py [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+    plt.show()
+
+def plot_cost(doc):
```
Comment (yarinshaked) [inline ex1/Utils.py:79]: add documentation

File: ex1/ex1.py [hunk 27]
```diff
@@ -0,0 +1,41 @@
+from LinearRegression import *
+from Utils import *
+
+
+if __name__ == "__main__":
```
Comment (yarinshaked) [inline ex1/ex1.py:5]: main function should be short, encapsulated in a different function

File: ex1/ex1.py [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+    # loading data
+    data_dir1 = "data\ex1data1.txt"
```
Comment (yarinshaked) [inline ex1/ex1.py:7]: should be command line argument

File: ex1/ex1.py [hunk 29]
```diff
@@ -0,0 +1,41 @@
...
+    data_dir2 = "data\ex1data2.txt"
```
Comment (yarinshaked) [inline ex1/ex1.py:8]: same here

**Review by yarinshaked — CHANGES_REQUESTED** [review]
Mazal tov, generally pretty good for the first time!

File: ex1/LinearRegression.py (hunk 3 above)
Comment (YairAmar) [reply ex1/LinearRegression.py]: https://www.python.org/dev/peps/pep-0257/
used this, which felt not very comprehensive...
not sure if i did fine

File: ex1/Utils.py (hunk 17 above)
Comment (YairAmar) [reply ex1/Utils.py]: because range returns some sort of an iterable object when i specifically want a list

**Review by yarinshaked — APPROVED** [review]
Mazal Tov, one step closer to not being a noob

**Pull request merged**